doc/continuous_integration/circleci.md

```diff
@@ -0,0 +1,104 @@
+
+* [CircleCI](https://circleci.com/)
+* [Navigation2 on CircleCI](https://circleci.com/gh/ros-planning/navigation2)
+
+For this particular CI, Docker is used to build and test the project within containers derived from images pulled from DockerHub, bootstrapping the CI with a development environment including pre configured dependencies and warm workspace build caches. View the accompanying DockerFile and DockerHub documentation for more info on this accompanying CI setup.
+
+* [Dockerfile](dockerfile.md)
+* [DockerHub](dockerhub.md)
+
+CircleCI is configured via the [config.yml](/.circleci/config.yml) yaml file within the `.circleci/` folder at the root of the GitHub repo.  The config file for this project is self-contained and thus densely structured, yet written in a functional style to remain DRY and modular. This keeps it easily generalizable for other ROS packages or for adjusting overlayed workspaces. Despite the anchors and references in yaml, the config file is best approached in reverse, from bottom to top in order of abstraction hierarchy, while reading this accompanied document. Further references on CircleCI configurations, such as syntax and structure can be found here:
+
+* [Writing YAML](https://circleci.com/docs/2.0/writing-yaml)
+* [Configuring CircleCI](https://circleci.com/docs/2.0/configuration-reference)
+
+## Workflows
+
+The CI config consists of three main [workflows](https://circleci.com/docs/2.0/configuration-reference/#workflows). One workflow is for checking PR events, essentially triggered by any pushed commits targeting the main branch, by building and testing the project both in release and debug mode for accurate performance benchmarking or generating test coverage results. The second is nightly cron scheduled to check the main branch, while additionally testing a matrix of supported RMW vendors not tested on normal PRs. This helps prioritize CI to quickly check new contributions, while simultaneously keeping tabs on the health of existing code. The third is another cron for updating CI image builds on DockerHub, and is scheduled to finish prior to the nightly workflow. This reduces the chance of updating CI images while a CI workflow is in progress.
+
+The order in which jobs are executed is conditional upon the completion of those it [requires](https://circleci.com/docs/2.0/configuration-reference/#requires), forming a conventional directed acyclic graph of dependent jobs. Independent jobs may of course be parallelized in the CI pipeline; so by splitting the build and test jobs in two, multiple test jobs, such as the matrix of RMW tests, may commence as soon as the dependent build job completes, avoiding unnecessarily re-building the same codebase for each RMW vendor.
```

Comment:
Be specific here, why are you talking about this? You should tell the reader why how you're using the `requires` in each of the 3 use cases you highlight in the paragraph above. Clearly one is for the 4 build/test jobs for the main CI situation. You abstractly say that it "may of course be parallelized in the CI pipeline", so tell them why you're bringing this up at all - you're leveraging that feature to speed up the pipelines by having many running jobs in parallel. Specifically go over that design for the 3 cases since this is the reference guide for our CircleCI setup. "We use this in the PR workflow by.... We use this in the nightly workflow by... etc"